docs: adding intro to sway (#55)
FuelLabs/docs-hub: docs/guides/docs/intro-to-sway/typescript-sdk.mdx

---
title: Typescript Frontend
category: Intro to Sway
parent:
  label: Guides
  link: /guides
---

## Building the Frontend

### Setup

Initialize a new React app with TypeScript in the same parent folder as your contract using the command below.

```shell

> npx create-react-app frontend --template typescript

Reviewer: I understand that the example was already generated using CRA. However, should we move forward into using vite?

Author: I think this is fine for now

```

Next, install the fuels Typescript and wallet SDKs in the frontend folder and generate types from your contract with `fuels typegen`.

```shell
cd frontend

> npm install fuels@0.60.0 @fuel-wallet/sdk@0.13.0 --save

Reviewer: Include  `@fuel-wallet/react`

npx fuels typegen -i ../contract/out/debug/*-abi.json -o ./src/contracts
```

Open the `src/App.tsx` file, and replace the boilerplate code with the template below:

<CodeImport
  file="../../examples/intro-to-sway/frontend/src/App.tsx"
  comment="fe_app_template"
  commentType="//"
  lang="tsx"
/>

Finally, copy and paste the CSS code below in your `App.css` file to add some simple styling.

<CodeImport
  file="../../examples/intro-to-sway/frontend/src/App.css"
  comment="fe_css"
  commentType="//"
  lang="tsx"
/>

### Connecting to the contract

At the top of the file, add your contract ID as a constant.

<CodeImport
  file="../../examples/intro-to-sway/frontend/src/App.tsx"
  comment="fe_contract_id"
  commentType="//"
  lang="tsx"
/>

> Next, create a new folder in the `src` folder called `hooks`, and copy and paste the `useFuel.tsx` and `useIsConnected.tsx` hooks from the example in this repo. You can also find them in the [offical wallet docs](https://wallet.fuel.network/docs/react/).

Reviewer: I would change this section, since the hooks can be found at @fuel-wallet/react.

More details [here](https://github.com/FuelLabs/fuels-wallet/blob/docs/fuel-wallet-react/packages/docs/docs/dev/getting-started.mdx#L84C17-L84C17).

Author: Yup this is outdated good catch

---
title: Typescript Frontend
category: Intro to Sway
parent:
  label: Guides
  link: /guides
---

## Building the Frontend

### Setup

Initialize a new React app with TypeScript in the same parent folder as your contract using the command below.

```shell
npx create-react-app frontend --template typescript
```

Next, install the fuels Typescript and wallet SDKs in the frontend folder and generate types from your contract with `fuels typegen`.

```shell
cd frontend

> npm install fuels @fuel-wallet/sdk

Reviewer 2: we need to specify versions for both of these like in the quickstart

npx fuels typegen -i ../contract/out/debug/*-abi.json -o ./src/contracts
```

In the `tsconfig.json` file, add the line below in the `compilerOptions` object to add the Fuel wallet type on the window object.

```json

> "types": ["@fuel-wallet/sdk"],

Reviewer 2: I don't think this is needed anymore

<CodeImport
  file="../../../intro-to-sway/frontend/src/App.tsx"
  comment="fe_items_contract"
  commentType="{/*"
  lang="tsx"
/>

Now, try listing an item to make sure this works. 
You should see the message `Item successfully listed!`.

### Show All Items

Next, let's create a new file called `AllItems.tsx` in the `components` folder.

Copy and paste the template code below for this component:

<CodeImport
  file="../../../intro-to-sway/frontend/src/components/AllItems.tsx"
  comment="fe_all_items_template"
  commentType="//"
  lang="tsx"
/>

Here we can get the item count to see how many items are listed, and then loop through each of them to get the item details.

First, let's create some state variables to store the number of items listed, an array of the item details, and the loading status.

<CodeImport
  file="../../../intro-to-sway/frontend/src/components/AllItems.tsx"
  comment="fe_all_items_state_variables"
  commentType="//"
  lang="tsx"
/>

Next, let's fetch the items in a `useEffect` hook.
Because these are read-only functions, we can simulate a dry-run of the transaction by using the `get` method instead of `call` so the user doesn't have to sign anything.

> <CodeImport

Reviewer 2: This is a change for the intro-to-sway repo, but there is a new method in the ts sdk to convert to a number, we should use this instead of my work-around `parseFloat(value.format()) * 1_000_000_000;`

Author: fixed here https://github.com/FuelLabs/intro-to-sway/pull/21
- [ ] update docs
- [ ] pull in new intro-to-sway here

```

Open the `src/App.tsx` file, and replace the boilerplate code with the template below:

<TestAction
id="fe-app-template"
action={{
  name: 'writeToFile',
  filepath: 'guides-testing/fuel-project/frontend/src/App.tsx'
}}
/>

<CodeImport
  file="../../examples/intro-to-sway/frontend/src/App.tsx"
  comment="fe_app_all"
  commentType="//"
  lang="tsx"
/>

Copy and paste the CSS code below in your `App.css` file to add some simple styling.

<TestAction
id="fe-css-template"
action={{
  name: 'writeToFile',
  filepath: 'guides-testing/fuel-project/frontend/src/App.css'
}}
/>

<CodeImport
  file="../../examples/intro-to-sway/frontend/src/App.css"
  comment="fe_css"
  commentType="//"
  lang="tsx"
/>

### Connecting to the contract

Wrap your components with `FuelProvider` from `@fuel-wallet/react` to enable Fuel's custom React hooks for wallet functionalities in `index.tsx`.

<TestAction
id="fe-index-all"
action={{
  name: 'writeToFile',
  filepath: 'guides-testing/fuel-project/frontend/src/index.tsx'
}}
/>

<CodeImport
  file="../../examples/intro-to-sway/frontend/src/index.tsx"
  comment="fe_index_all"
  commentType="//"
  lang="tsx"
/>

> At the top of the file, add your contract ID that you had earlier as a constant.

Reviewer 2: This is out of place after editing the `index.tsx` file. I think it makes sense to edit `index.tsx` and then `App.tsx`.

<CodeImport
  file="../../examples/intro-to-sway/frontend/src/App.tsx"
  comment="fe_contract_id"
  commentType="//"
  lang="tsx"
/>

> Webhooks from the Fuel Wallet package are used in order to connect our wallet to the dapp. In the `App` function, we can call these hooks like this:

Reviewer 2: ```suggestion
React hooks from the Fuel Wallet package are used in order to connect our wallet to the dapp. In the `App` function, we can call these hooks like this:
```

<CodeImport
  file="../../examples/intro-to-sway/frontend/src/App.tsx"
  comment="fe_call_hooks"
  commentType="//"
  lang="tsx"
/>

> The state variable called `wallets` with the `useState` hook will have the type `WalletLocked`.

Reviewer 2: ```suggestion
The `wallet` variable from the `useWallet` hook will have the type `FuelWalletLocked`.
```
